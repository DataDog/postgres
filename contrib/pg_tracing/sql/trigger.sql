-- Only trace queries with sample flag
SET pg_tracing.sample_rate = 0.0;
SET pg_tracing.caller_sample_rate = 1.0;

-- Create test tables
CREATE TABLE Employee (
    EmployeeId INT NOT NULL,
    LastName VARCHAR(20) NOT NULL,
    CONSTRAINT PK_Employee PRIMARY KEY (EmployeeId));
CREATE TABLE Employee_Audit (
    EmployeeId INT NOT NULL,
    LastName VARCHAR(20) NOT NULL,
    EmpAdditionTime VARCHAR(20) NOT NULL);

-- Create test trigger
CREATE OR REPLACE FUNCTION employee_insert_trigger_fnc()
  RETURNS trigger AS
$$
BEGIN
    INSERT INTO Employee_Audit (EmployeeId, LastName, EmpAdditionTime)
         VALUES(NEW.EmployeeId,NEW.LastName,current_date);
RETURN NEW;
END;
$$
LANGUAGE plpgsql;

-- Hook the trigger twice
CREATE TRIGGER employee_insert_trigger
  AFTER INSERT
  ON Employee
  FOR EACH ROW
  EXECUTE PROCEDURE employee_insert_trigger_fnc();

CREATE TRIGGER employee_insert_trigger_2
  AFTER INSERT
  ON Employee
  FOR EACH ROW
  EXECUTE PROCEDURE employee_insert_trigger_fnc();


-- Call update
/*dddbs='postgres.db',traceparent='00-00000000000000000000000000000001-0000000000000001-01'*/ INSERT INTO Employee VALUES(10,'Adams');


--
-- +--------+---------------------------------------------------------------------------------------------------------------------------------------+
-- |A: Parse| B: INSERT INTO Employee                                                                                                               |
-- +--------++----------+-+-------------------------+---------------------------------------------------------------------------------------------+-+
--           |C: Planner| |D: ExecutorRun           | G: ExecutorFinish                                                                           |
--           +----------+ ++-----------------------++--+--------+-------------------------------------------+-+---------------------------------+-+
--                         |E: Insert into employee|   |H: Parse|  I: INSERT INTO Employee_Audit...         | |N: INSERT INTO Employee_Audit...|
--                         ++---------+------------+   +--------+-+----------++----------------------------++ +--------------------------------+
--                          |F: Result|                           |J: Planner||K: ExecutorRun              |  |O: ExecutorRun |
--                          +---------+                           +----------+++---------------------------+  ++--------------+
--                                                                             |L: Insert on employee audit|   | ...          |
--                                                                             +-+---------+---------------+   +--------------+
--                                                                               |M: Result|
--                                                                               +---------+

-- Gather span_id, span start and span end of call with triggers
SELECT span_id AS span_a_id,
        get_span_start(span_start) as span_a_start,
        get_span_end(span_start) as span_a_end
		from pg_tracing_peek_spans where parent_id='0000000000000001' and span_type='Parse' \gset
SELECT span_id AS span_b_id,
        get_span_start(span_start) as span_b_start,
        get_span_end(span_start) as span_b_end
		from pg_tracing_peek_spans where parent_id='0000000000000001' and span_type!='Parse' \gset
SELECT span_id AS span_c_id,
        get_span_start(span_start) as span_c_start,
        get_span_end(span_start) as span_c_end
		from pg_tracing_peek_spans where parent_id=:'span_b_id' and span_type='Planner' \gset
SELECT span_id AS span_d_id,
        get_span_start(span_start) as span_d_start,
        get_span_end(span_start) as span_d_end
		from pg_tracing_peek_spans where parent_id=:'span_b_id' and span_type='Executor' and span_operation='ExecutorRun' \gset
SELECT span_id AS span_e_id,
        get_span_start(span_start) as span_e_start,
        get_span_end(span_start) as span_e_end
		from pg_tracing_peek_spans where parent_id=:'span_d_id' and span_type='Insert' \gset
SELECT span_id AS span_f_id,
        get_span_start(span_start) as span_f_start,
        get_span_end(span_start) as span_f_end
		from pg_tracing_peek_spans where parent_id=:'span_e_id' and span_type='Result' \gset

-- Executor Finish, first trigger
SELECT span_id AS span_g_id,
        get_span_start(span_start) as span_g_start,
        get_span_end(span_start) as span_g_end
		from pg_tracing_peek_spans where parent_id=:'span_b_id' and span_type='Executor' and span_operation='ExecutorFinish' \gset
SELECT span_id AS span_h_id,
        get_span_start(span_start) as span_h_start,
        get_span_end(span_start) as span_h_end
		from pg_tracing_peek_spans where parent_id=:'span_g_id' and span_type='Parse' \gset
SELECT span_id AS span_i_id,
        get_span_start(span_start) as span_i_start,
        get_span_end(span_start) as span_i_end
		from pg_tracing_peek_spans where parent_id=:'span_g_id' and span_type='Insert query' limit 1 \gset
SELECT span_id AS span_j_id,
        get_span_start(span_start) as span_j_start,
        get_span_end(span_start) as span_j_end
		from pg_tracing_peek_spans where parent_id=:'span_i_id' and span_type='Planner' \gset
SELECT span_id AS span_k_id,
        get_span_start(span_start) as span_k_start,
        get_span_end(span_start) as span_k_end
		from pg_tracing_peek_spans where parent_id=:'span_i_id' and span_type='Executor' and span_operation='ExecutorRun' \gset
SELECT span_id AS span_l_id,
        get_span_start(span_start) as span_l_start,
        get_span_end(span_start) as span_l_end
		from pg_tracing_peek_spans where parent_id=:'span_k_id' and span_type='Insert' \gset
SELECT span_id AS span_m_id,
        get_span_start(span_start) as span_m_start,
        get_span_end(span_start) as span_m_end
		from pg_tracing_peek_spans where parent_id=:'span_l_id' and span_type='Result' \gset

-- Executor Finish, second trigger
SELECT span_id AS span_n_id,
        get_span_start(span_start) as span_n_start,
        get_span_end(span_start) as span_n_end
        from pg_tracing_peek_spans where parent_id=:'span_g_id' and span_type='Insert query' limit 1 offset 1 \gset
SELECT span_id AS span_o_id,
        get_span_start(span_start) as span_o_start,
        get_span_end(span_start) as span_o_end
        from pg_tracing_peek_spans where parent_id=:'span_n_id' and span_type='Executor' and span_operation='ExecutorRun' \gset

-- Check that spans' start and end are within expection
-- Root span
SELECT :span_a_end <= :span_b_start as parse_before_root_span;
-- Planner
SELECT :span_c_start >= :span_b_start as planner_starts_after_root_span;
-- ExecutorRun
SELECT :span_d_start >= :span_c_end as executor_run_starts_after_planner,
       :span_d_end <= :span_g_start as executor_run_ends_before_finish;
-- First Insert
SELECT :span_e_start >= :span_d_start as insert_starts_after_executor_run,
       :span_e_start <= :span_f_start as insert_ends_before_executor_run,
       :span_e_end >= :span_f_end as insert_starts_before_child,
       :span_e_end <= :span_d_end as insert_ends_after_child;
-- ExecutorFinish
SELECT :span_g_start <= :span_i_start as executor_finish_starts_before_child,
       :span_g_end >= :span_o_end as executor_finish_ends_after_child;
-- First trigger
SELECT :span_h_end <= :span_i_start as parse_ends_before_first_trigger,
       :span_i_end >= :span_n_end as first_trigger_ends_after_child;
-- Second trigger
SELECT :span_o_start >= :span_i_end  as second_trigger_starts_after_first,
       :span_o_end >= :span_o_end as second_trigger_ends_after_child,
       :span_o_end <= :span_g_end as second_trigger_ends_before_executorfinish_end,
       :span_n_end <= :span_g_end as second_trigger_ends_before_root_executorfinish_end;

-- Check span_operation
SELECT span_type, span_operation, lvl from peek_ordered_spans where trace_id='00000000000000000000000000000001';

-- Check count of query_id
SELECT count(distinct query_id) from pg_tracing_consume_spans where trace_id='00000000000000000000000000000001';

-- Second call, parse should not be present anymore
/*dddbs='postgres.db',traceparent='00-00000000000000000000000000000002-0000000000000002-01'*/ INSERT INTO Employee VALUES(11,'Adams');

SELECT span_id AS span_b_id,
        get_span_start(span_start) as span_b_start,
        get_span_end(span_start) as span_b_end
		from pg_tracing_peek_spans where parent_id='0000000000000002' and span_type!='Parse' \gset
SELECT span_id AS span_d_id,
        get_span_start(span_start) as span_d_start,
        get_span_end(span_start) as span_d_end
		from pg_tracing_peek_spans where parent_id=:'span_b_id' and span_type='Executor' and span_operation='ExecutorRun' \gset
SELECT span_id AS span_g_id,
        get_span_start(span_start) as span_g_start,
        get_span_end(span_start) as span_g_end
		from pg_tracing_peek_spans where parent_id=:'span_b_id' and span_type='Executor' and span_operation='ExecutorFinish' \gset

-- Check we don't have Parse span anymore
SELECT count(*) = 0 from pg_tracing_peek_spans where parent_id=:'span_g_id' and span_type='Parse';
-- Insert should be attached to the correct parent
SELECT count(*) = 1 from pg_tracing_peek_spans where parent_id=:'span_g_id' and span_type='Insert' limit 1 \gset

-- Check that ExecutorRun ends before ExecutorFinish start
SELECT :span_d_end <= :span_g_start as finish_start_after_run;

-- Check generated spans for the second call with trigger
SELECT span_type, span_operation, lvl from peek_ordered_spans where trace_id='00000000000000000000000000000002';

-- Check count of query_id
SELECT count(distinct query_id) from pg_tracing_consume_spans where trace_id='00000000000000000000000000000002';

-- Test table to test before trigger
CREATE TABLE before_trigger_table (a int);

-- Create before trigger function
CREATE OR REPLACE FUNCTION before_trigger_fnc()
  RETURNS trigger AS
$$
BEGIN
    PERFORM 'SELECT 1';
RETURN NEW;
END;
$$
LANGUAGE plpgsql;

-- Hook the before trigger
CREATE TRIGGER employee_insert_trigger
  BEFORE INSERT
  ON before_trigger_table
  FOR EACH ROW
  EXECUTE PROCEDURE before_trigger_fnc();

-- Test explain on a before trigger
-- This will go through ExecutorEnd without any parent executor run
/*dddbs='postgres.db',traceparent='00-fed00000000000000000000000000001-0000000000000003-01'*/ explain INSERT INTO before_trigger_table VALUES(10);
SELECT span_type, span_operation, lvl from peek_ordered_spans where trace_id='fed00000000000000000000000000001';

-- Call before trigger update
/*dddbs='postgres.db',traceparent='00-00000000000000000000000000000004-0000000000000004-01'*/ INSERT INTO before_trigger_table VALUES(10);
SELECT span_type, span_operation, lvl from peek_ordered_spans where trace_id='00000000000000000000000000000004';

-- Check count of query_id
SELECT count(distinct query_id) from pg_tracing_consume_spans;
