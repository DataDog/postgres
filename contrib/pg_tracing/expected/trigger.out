-- Only trace queries with sample flag
SET pg_tracing.sample_rate = 0.0;
SET pg_tracing.caller_sample_rate = 1.0;
-- Create test tables
CREATE TABLE Employee (
    EmployeeId INT NOT NULL,
    LastName VARCHAR(20) NOT NULL,
    CONSTRAINT PK_Employee PRIMARY KEY (EmployeeId));
CREATE TABLE Employee_Audit (
    EmployeeId INT NOT NULL,
    LastName VARCHAR(20) NOT NULL,
    EmpAdditionTime VARCHAR(20) NOT NULL);
-- Create test trigger
CREATE OR REPLACE FUNCTION employee_insert_trigger_fnc()
  RETURNS trigger AS
$$
BEGIN
    INSERT INTO Employee_Audit (EmployeeId, LastName, EmpAdditionTime)
         VALUES(NEW.EmployeeId,NEW.LastName,current_date);
RETURN NEW;
END;
$$
LANGUAGE plpgsql;
-- Hook the trigger twice
CREATE TRIGGER employee_insert_trigger
  AFTER INSERT
  ON Employee
  FOR EACH ROW
  EXECUTE PROCEDURE employee_insert_trigger_fnc();
CREATE TRIGGER employee_insert_trigger_2
  AFTER INSERT
  ON Employee
  FOR EACH ROW
  EXECUTE PROCEDURE employee_insert_trigger_fnc();
-- Call update
/*dddbs='postgres.db',traceparent='00-00000000000000000000000000000001-0000000000000001-01'*/ INSERT INTO Employee VALUES(10,'Adams');
-- +-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
-- | A: INSERT INTO Employee                                                                                                                                                                       |
-- +-+--------+-+----------+-+--------------------------+------------------------------------------------------------------------------------------------------------------------+---------------+-+
--   |B: Parse| |C: Planner| |D: ExecutorRun            | G: ExecutorFinish                                                                                                      |R: ExecutorEnd |
--   +--------+ +----------+ ++-----------------------+-+--+----------------------------------------------------------------------+--+-------------------------------------------+---------------+
--                            |E: Insert into employee|    |H: INSERT INTO Employee_Audit...                                      |  |O: INSERT INTO Employee_Audit...           |
--                            ++---------+------------+    ++--------++----------++----------------------------+---+--------------+  ++-----+---------------+-----+--------------+
--                             |F: Result|                  |I: Parse||J: Planner||K: ExecutorRun              |...|N: ExecutorEnd|   | ... |P: ExecutorRun | ... |Q: ExecutorEnd|
--                             +---------+                  +--------++----------+++---------------------------+   +--------------+   +-----++--------------+     +--------------+
--                                                                                 |L: Insert on employee audit|                             | ...          |
--                                                                                 +-+---------+---------------+                             +--------------+
--                                                                                   |M: Result|
--                                                                                   +---------+
-- Gather span_id, span start and span end of call with triggers
SELECT span_id AS span_a_id,
        get_span_start_ns(span_start, span_start_ns) as span_a_start,
        get_span_end_ns(span_start, span_start_ns, duration) as span_a_end
		from pg_tracing_peek_spans where parent_id=1 and name!='Parse' \gset
SELECT span_id AS span_b_id,
        get_span_start_ns(span_start, span_start_ns) as span_b_start,
        get_span_end_ns(span_start, span_start_ns, duration) as span_b_end
		from pg_tracing_peek_spans where parent_id=:span_a_id and name='Parse' \gset
SELECT span_id AS span_c_id,
        get_span_start_ns(span_start, span_start_ns) as span_c_start,
        get_span_end_ns(span_start, span_start_ns, duration) as span_c_end
		from pg_tracing_peek_spans where parent_id=:span_a_id and name='Planner' \gset
SELECT span_id AS span_d_id,
        get_span_start_ns(span_start, span_start_ns) as span_d_start,
        get_span_end_ns(span_start, span_start_ns, duration) as span_d_end
		from pg_tracing_peek_spans where parent_id=:span_a_id and name='Executor' and resource='Run' \gset
SELECT span_id AS span_e_id,
        get_span_start_ns(span_start, span_start_ns) as span_e_start,
        get_span_end_ns(span_start, span_start_ns, duration) as span_e_end
		from pg_tracing_peek_spans where parent_id=:span_d_id and name='Insert' \gset
SELECT span_id AS span_f_id,
        get_span_start_ns(span_start, span_start_ns) as span_f_start,
        get_span_end_ns(span_start, span_start_ns, duration) as span_f_end
		from pg_tracing_peek_spans where parent_id=:span_e_id and name='Result' \gset
-- Executor Finish, first trigger
SELECT span_id AS span_g_id,
        get_span_start_ns(span_start, span_start_ns) as span_g_start,
        get_span_end_ns(span_start, span_start_ns, duration) as span_g_end
		from pg_tracing_peek_spans where parent_id=:span_a_id and name='Executor' and resource='Finish' \gset
SELECT span_id AS span_h_id,
        get_span_start_ns(span_start, span_start_ns) as span_h_start,
        get_span_end_ns(span_start, span_start_ns, duration) as span_h_end
		from pg_tracing_peek_spans where parent_id=:span_g_id and name='Insert' limit 1 \gset
SELECT span_id AS span_i_id,
        get_span_start_ns(span_start, span_start_ns) as span_i_start,
        get_span_end_ns(span_start, span_start_ns, duration) as span_i_end
		from pg_tracing_peek_spans where parent_id=:span_h_id and name='Parse' \gset
SELECT span_id AS span_j_id,
        get_span_start_ns(span_start, span_start_ns) as span_j_start,
        get_span_end_ns(span_start, span_start_ns, duration) as span_j_end
		from pg_tracing_peek_spans where parent_id=:span_h_id and name='Planner' \gset
SELECT span_id AS span_k_id,
        get_span_start_ns(span_start, span_start_ns) as span_k_start,
        get_span_end_ns(span_start, span_start_ns, duration) as span_k_end
		from pg_tracing_peek_spans where parent_id=:span_h_id and name='Executor' and resource='Run' \gset
SELECT span_id AS span_l_id,
        get_span_start_ns(span_start, span_start_ns) as span_l_start,
        get_span_end_ns(span_start, span_start_ns, duration) as span_l_end
		from pg_tracing_peek_spans where parent_id=:span_k_id and name='Insert' \gset
SELECT span_id AS span_m_id,
        get_span_start_ns(span_start, span_start_ns) as span_m_start,
        get_span_end_ns(span_start, span_start_ns, duration) as span_m_end
		from pg_tracing_peek_spans where parent_id=:span_l_id and name='Result' \gset
SELECT span_id AS span_n_id,
        get_span_start_ns(span_start, span_start_ns) as span_n_start,
        get_span_end_ns(span_start, span_start_ns, duration) as span_n_end
        from pg_tracing_peek_spans where parent_id=:span_h_id and name='Executor' and resource='End' \gset
-- Executor Finish, second trigger
SELECT span_id AS span_o_id,
        get_span_start_ns(span_start, span_start_ns) as span_o_start,
        get_span_end_ns(span_start, span_start_ns, duration) as span_o_end
        from pg_tracing_peek_spans where parent_id=:span_g_id and name='Insert' limit 1 offset 1 \gset
SELECT span_id AS span_p_id,
        get_span_start_ns(span_start, span_start_ns) as span_p_start,
        get_span_end_ns(span_start, span_start_ns, duration) as span_p_end
        from pg_tracing_peek_spans where parent_id=:span_o_id and name='Executor' and resource='Run' \gset
SELECT span_id AS span_q_id,
        get_span_start_ns(span_start, span_start_ns) as span_q_start,
        get_span_end_ns(span_start, span_start_ns, duration) as span_q_end
        from pg_tracing_peek_spans where parent_id=:span_o_id and name='Executor' and resource='End' \gset
-- End
SELECT span_id AS span_r_id,
        get_span_start_ns(span_start, span_start_ns) as span_r_start,
        get_span_end_ns(span_start, span_start_ns, duration) as span_r_end
        from pg_tracing_peek_spans where parent_id=:span_a_id and name='Executor' and resource='End' \gset
-- Check that spans' start and end are within expection
SELECT
-- Root span
        :span_a_start <= :span_b_start as root_span_starts_first,
-- Planner
       :span_c_start >= :span_b_end as planner_after_parse,
-- ExecutorRun
       :span_d_start >= :span_c_end as executor_run_starts_after_planner,
       :span_d_end <= :span_g_start as executor_run_ends_before_finish,
-- First Insert
       :span_e_start >= :span_d_start as insert_starts_after_executor_run,
       :span_e_start <= :span_f_start as insert_ends_before_executor_run,
       :span_e_end >= :span_f_end as insert_starts_before_child,
       :span_e_end <= :span_d_end as insert_ends_after_child,
-- ExecutorFinish
       :span_g_start <= :span_h_start as executor_finish_starts_before_child,
       :span_g_end >= :span_o_end as executor_finish_ends_after_child,
-- First trigger
       :span_h_start <= :span_i_start as first_trigger_starts_before_child,
       :span_h_end >= :span_n_end as first_trigger_ends_after_child,
-- Second trigger
       :span_o_start >= :span_h_end  as second_trigger_starts_after_first,
       :span_o_end >= :span_q_end as second_trigger_ends_after_child,
       :span_q_end <= :span_o_start as second_trigger_ends_before_executor_end,
       :span_o_end <= :span_r_start as second_trigger_ends_before_root_executor_end,
-- Root ExecutorEnd
       :span_r_start >= :span_g_end as executor_end_start_after_executor_finish,
       :span_r_end <= :span_a_end as executor_end_ends_before_root;
 root_span_starts_first | planner_after_parse | executor_run_starts_after_planner | executor_run_ends_before_finish | insert_starts_after_executor_run | insert_ends_before_executor_run | insert_starts_before_child | insert_ends_after_child | executor_finish_starts_before_child | executor_finish_ends_after_child | first_trigger_starts_before_child | first_trigger_ends_after_child | second_trigger_starts_after_first | second_trigger_ends_after_child | second_trigger_ends_before_executor_end | second_trigger_ends_before_root_executor_end | executor_end_start_after_executor_finish | executor_end_ends_before_root 
------------------------+---------------------+-----------------------------------+---------------------------------+----------------------------------+---------------------------------+----------------------------+-------------------------+-------------------------------------+----------------------------------+-----------------------------------+--------------------------------+-----------------------------------+---------------------------------+-----------------------------------------+----------------------------------------------+------------------------------------------+-------------------------------
 t                      | t                   | t                                 | t                               | t                                | t                               | t                          | t                       | t                                   | t                                | t                                 | t                              | t                                 | t                               | f                                       | t                                            | t                                        | t
(1 row)

-- Check resource and query id
SELECT name, resource, query_id from pg_tracing_consume_spans where trace_id=1 order by span_start, span_start_ns, resource;
    name    |                                                      resource                                                       |       query_id       
------------+---------------------------------------------------------------------------------------------------------------------+----------------------
 Insert     | INSERT INTO Employee VALUES($1,$2);                                                                                 |  1426892027161297518
 Parse      | Parse                                                                                                               |  1426892027161297518
 Post Parse | Post Parse                                                                                                          |  1426892027161297518
 Planner    | Planner                                                                                                             |  1426892027161297518
 Executor   | Start                                                                                                               |  1426892027161297518
 Executor   | Run                                                                                                                 |  1426892027161297518
 Insert     | Insert on employee                                                                                                  |  1426892027161297518
 Result     | Result                                                                                                              |  1426892027161297518
 Executor   | Finish                                                                                                              |  1426892027161297518
 Insert     | INSERT INTO Employee_Audit (EmployeeId, LastName, EmpAdditionTime) VALUES(NEW.EmployeeId,NEW.LastName,current_date) | -4246438446062883031
 Parse      | Parse                                                                                                               | -4246438446062883031
 Post Parse | Post Parse                                                                                                          | -4246438446062883031
 Planner    | Planner                                                                                                             | -4246438446062883031
 Executor   | Start                                                                                                               | -4246438446062883031
 Executor   | Run                                                                                                                 | -4246438446062883031
 Insert     | Insert on employee_audit                                                                                            | -4246438446062883031
 Result     | Result                                                                                                              | -4246438446062883031
 Executor   | Finish                                                                                                              | -4246438446062883031
 Executor   | End                                                                                                                 | -4246438446062883031
 Insert     | INSERT INTO Employee_Audit (EmployeeId, LastName, EmpAdditionTime) VALUES(NEW.EmployeeId,NEW.LastName,current_date) | -4246438446062883031
 Parse      | Parse                                                                                                               | -4246438446062883031
 Post Parse | Post Parse                                                                                                          | -4246438446062883031
 Planner    | Planner                                                                                                             | -4246438446062883031
 Executor   | Start                                                                                                               | -4246438446062883031
 Executor   | Run                                                                                                                 | -4246438446062883031
 Insert     | Insert on employee_audit                                                                                            | -4246438446062883031
 Result     | Result                                                                                                              | -4246438446062883031
 Executor   | Finish                                                                                                              | -4246438446062883031
 Executor   | End                                                                                                                 | -4246438446062883031
 Executor   | End                                                                                                                 |  1426892027161297518
(30 rows)

